Simplify category form

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Simplifies category form and refactors category saving, by adding more abstract classes. This makes sure that there is less repeated code between root and normal category management. Root category form type was already abstracted so there is little reason why form handlers shouldn't be abstracted as well.
| Type?         | refacto
| Category?     | BO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Part of #16482
| How to test?  | See below

## How to test

Simplifying Sell -> Catalog -> Category. Everything must look the same.

Please test the "add menu thumbnail images" capacity: add multiple, remove, and there is a limit (3) above which **the input is disabled**.

The description should have clean HTML(no javascript). Some fields can't exceed a certain number of characters meta title and meta keywords: 255, meta description: 512, name and link rewrite: 128 characters

## :notebook: BC Break
Backwards compatibility break introduced because AbstractCategoryType has changed parameters. This means that if any module extends this type(or any type that extends it) that module might crash upon updating to PS version where this update is present.
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/21243)
<!-- Reviewable:end -->

JevgenijVisockij committed:
Changed the way disable_menu_thumbnails_upload is used in form.

JevgenijVisockij committed:
Fixed getting an exception when menu thumbnails is at maximum capacity

JevgenijVisockij committed:
Removing help from description as it's not correct

JevgenijVisockij committed:
Added comment to make changes clearer

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Changed help text for category link_rewerite to the one that makes more sense

JevgenijVisockij committed:
Removed no longer needed help text

JevgenijVisockij committed:
Fixed problems with root category

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Fixed category images not showing up

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php

Added dot constraint message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php

Added dot to constraint message.

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php

Added dot to constraint message.

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/Core/Domain/Category/SeoSettings.php

Added public to constant definition

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Added label for parent category

JevgenijVisockij committed:
Added maximum values to certain category fields

# Conflicts:
#	src/Core/Domain/Category/SeoSettings.php

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Shortended the syntax

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Using short if syntax

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Added bugfix in case you are adding category from root

JevgenijVisockij committed:
CS fixer fix

matks committed:
Apply suggestions from code review

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Replaced self with static

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Checking isset for menu_thumbnails_images

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Checking for isset for menu thumbnail images

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Using as decimal in sprintf instead of string

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Making MAX_TITLE_LENGTH public

JevgenijVisockij committed:
Making constraint for max title length instead of SeoSettings max title length

JevgenijVisockij committed:
Changed from root category to home category

JevgenijVisockij committed:
Adding test for getAvailableKeys function

JevgenijVisockij committed:
Moved most of the category handler code to abstract handler

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij committed:
Implemented feedback for getAvailableKeys function and its test

JevgenijVisockij committed:
Removed constant changing

JevgenijVisockij committed:
Added abstract functions

JevgenijVisockij committed:
Fixed abstract category handler

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Added int typehint for categoryId

JevgenijVisockij committed:
Changed categoryId so it get's converted to int earleir

JevgenijVisockij committed:
Apply suggestions from code review

Adding typehints

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Update src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php

Adding strict types

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Changed return types so that most params can be null

JevgenijVisockij committed:
made most params in add category command nullable

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Fixed issue with form not saving due to images when it should

JevgenijVisockij committed:
Removed HTML column since it's added in another PR

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij committed:
Fixed ESLint problem

JevgenijVisockij committed:
Update src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php

Adding array map to make sure all ids are int

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Update src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php

Removed uneeded checking for bool

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Removed no lnoger needed check

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig

Fixed twig forms url

Co-authored-by: atomiix <tbaccelli@gmail.com>
# Conflicts:
#	templates/bundles/PrestaShopBundle/Admin/Sell/Catalog/Categories/Blocks/form.html.twig

JevgenijVisockij committed:
Fixed help link not working

# Conflicts:
#	src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php

JevgenijVisockij committed:
Added back final

sowbiba committed:
Use internal annotation for final classes + add abstract keyword

JevgenijVisockij committed:
Adding back HTML column

l-delin commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -886,6 +910,16 @@ private function getErrorMessages()
             ),
+            UploadedImageConstraintException::class => [
+                UploadedImageConstraintException::EXCEEDED_SIZE => $this->trans(
+                    'Max file size allowed is "%d" bytes.', 'Admin.Notifications.Error',
```
```suggestion
                    'Max file size allowed is %d bytes.', 'Admin.Notifications.Error',
```

JevgenijVisockij committed:
Added proper error throwing for multiple images in menu thumbnails

# Conflicts:
#	src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php

JevgenijVisockij committed:
Added abstract root command to allow for typehinting

# Conflicts:
#	src/Core/Domain/Category/Command/AddCategoryCommand.php
#	src/Core/Domain/Category/Command/AddRootCategoryCommand.php
#	src/Core/Domain/Category/Command/EditCategoryCommand.php
#	src/Core/Domain/Category/Command/EditRootCategoryCommand.php

JevgenijVisockij committed:
Adding localized additional descriptions to abstract classes

JevgenijVisockij committed:
CS Fixer fixes

JevgenijVisockij committed:
CS Fixer fix

JevgenijVisockij committed:
Fixing design for category form

l-delin commented on src/Core/Form/IdentifiableObject/DataHandler/AbstractCategoryFormDataHandler.php:
```diff
@@ -0,0 +1,187 @@
+        $availableKeys = $this->getAvailableKeys($categoryId);
+
+        if (isset($data['menu_thumbnail_images']) && count($data['menu_thumbnail_images']) > count($availableKeys)) {
+            throw new MenuThumbnailsLimitException(sprintf('Maximum number of menu thumbnails was reached for category "%d"', $categoryId));
```
```suggestion
            throw new MenuThumbnailsLimitException(sprintf('The maximum number of menu thumbnails has been reached for the %d category', $categoryId));
```

l-delin commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -135,19 +167,37 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('cover_image', FileType::class, [
+                'label' => $this->trans('Category cover image', 'Admin.Catalog.Feature'),
+                'help' => $this->trans('This is the main image for your category, displayed in the category page. The category description will overlap this image and appear in its top-left corner.', 'Admin.Catalog.Help'),
```
```suggestion
                'help' => $this->trans('This is the cover image for your category: it will be displayed on the category\'s page. The description will appear in its top-left corner.', 'Admin.Catalog.Help'),
```

l-delin commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -211,36 +263,67 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('meta_keyword', TranslatableType::class, [
+                'label' => $this->trans('Meta keywords', 'Admin.Global'),
+                'help' => $this->trans('To add tags, click in the field, write something, and then press the "Enter" key.', 'Admin.Shopparameters.Help')
```
```suggestion
                'help' => $this->trans('To add tags, click in the field, write something, and then press the "Enter" key.', 'Admin.Shopparameters.Help')
```

I think this help text needs to be removed entirely. We could add a tooltip that explains Meta keywords next to it, e.g.: _"Search engines use meta keywords to index pages. Only you can see them."_

Ping @PrestaShop/ux-team

ElodieOS replied:
Hi @l-delin!

It's a great idea to add a tooltip that explains meta keywords but I wouldn't remove the help text, as we have specific behavior here. The merchant can put a space in the tag (this is not always the case). So maybe the user won't understand that to validate his tag he has to press "enter" key and not "space" key. But we can cut the sentence with : "To add tags, press the 'enter' key. Invalid characters: <>;=#{}

Wdyt?

l-delin replied:
Hello @ElodieOS 🙌

How about: "To add tags, press the 'enter' key. You can also use the 'comma' key. Invalid characters: <>;=#{}" ?

ElodieOS replied:
Hello @l-delin ! 
Yes perfect 👌

l-delin commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -211,36 +263,67 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('link_rewrite', TranslatableType::class, [
+                'label' => $this->trans('Friendly URL', 'Admin.Global'),
+                'help' => $this->trans('Unless the \'Accented URL\' option is enabled (in Shop parameters > Traffic & SEO), only letters, numbers, underscores (_), and hyphens (-) are allowed.', 'Admin.Catalog.Help'),
```
```suggestion
                'help' => $this->trans('Allowed characters: letters, numbers, underscores (_) and hyphens (-). To allow more characters, enable the \'Accented URL\' feature in Shop Parameters > Traffic & SEO.', 'Admin.Catalog.Help'),
```

l-delin commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -211,36 +263,67 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('group_association', MaterialChoiceTableType::class, [
+                'label' => $this->trans('Group access', 'Admin.Catalog.Feature'),
+                'help' => $this->trans('Mark all of the customer groups which you would like to have access to this category.', 'Admin.Catalog.Help'),
```
```suggestion
                'help' => $this->trans('Select the customer groups which will have access to this category.', 'Admin.Catalog.Help'),
```

l-delin commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -135,19 +167,37 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('active', SwitchType::class, [
+                'label' => $this->trans('Enabled', 'Admin.Global'),
+                'help' => $this->trans(
+                        'If you want a category to appear in the menu of your shop, go to [1]Modules > Module Manager[/1] and configure your menu module.',
```
```suggestion
                        'If you want a category to appear in your store's menu, configure your menu module in [1]Modules > Module Manager[/1].',
```

l-delin commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -135,19 +167,37 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('thumbnail_image', FileType::class, [
+                'label' => $this->trans('Category thumbnail', 'Admin.Catalog.Feature'),
+                'help' => $this->trans('Displays a small image in the parent category\'s page, if the theme allows it.', 'Admin.Catalog.Help'),
```
```suggestion
                'help' => $this->trans('It will display a thumbnail on the parent category\'s page, if the theme allows it.', 'Admin.Catalog.Help'),
```

l-delin commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -135,19 +167,37 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('menu_thumbnail_images', FileType::class, [
+                'label' => $this->trans('Menu thumbnails', 'Admin.Catalog.Feature'),
+                'help' => $this->trans('The category thumbnail appears in the menu as a small image representing the category, if the theme allows it.', 'Admin.Catalog.Help'),
```
```suggestion
                'help' => $this->trans('It will display a thumbnail representing the category in the menu, if the theme allows it.', 'Admin.Catalog.Help'),
```

l-delin reviewed (commented):
The wording needs some editing done, thanks :-)

l-delin commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -211,36 +263,67 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('meta_keyword', TranslatableType::class, [
+                'label' => $this->trans('Meta keywords', 'Admin.Global'),
+                'help' => $this->trans('To add tags, click in the field, write something, and then press the "Enter" key.', 'Admin.Shopparameters.Help')
```
```suggestion
                'help' => $this->trans('To add tags, press the 'enter' key. You can also use the 'comma' key. Invalid characters: <>;=#{}', 'Admin.Shopparameters.Help')
```

l-delin reviewed (commented):
edited wording for clarity :)

JevgenijVisockij committed:
Apply suggestions from code review

Wording improvements

Co-authored-by: l-delin <113511191+l-delin@users.noreply.github.com>

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij committed:
Fixing category translation

JevgenijVisockij committed:
Changed from escape to raw

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/Content/data.html.twig

JevgenijVisockij committed:
Added new "HTML" column

# Conflicts:
#	src/Core/Grid/Column/Type/Common/HtmlColumn.php

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/Content/data.html.twig

JevgenijVisockij committed:
Added TextWithRecommendedLengthCounter component

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.js
#	admin-dev/themes/new-theme/js/pages/category/index.js

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.js
#	admin-dev/themes/new-theme/js/pages/category/index.ts

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.ts

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.ts

JevgenijVisockij committed:
Removing uneeded spaces

zuk3975 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,317 @@
+/**
+ * Class EditRootCategoryCommand edits given root category.
+ */
+abstract class AbstractEditCategoryCommand
```
seems like you had some similar ideas as I did, but @jolelievre might not be up to this hehe :smile: 
https://github.com/PrestaShop/PrestaShop/pull/30231#discussion_r1025048099

jolelievre replied:
Indeed 😅 Although I do understand the need and the motivation here, but each command is supposed to be independent so they shouldn't be coupled by extension even if it's an abstract class (although it's less a problem with an abstract class maybe).

But the idea is that modifying the parent changes both commands and both contracts, which you can't assume to be true at all time. So to avoid some unexpected modification on unrelated command it is probably safer to keep commands independent. But I'm still not fully convinced of either choices... maybe we should discuss this at the next standup?

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -34,84 +34,35 @@
-        </div>
-      </div>
-
+      {{ form_row(categoryForm.name) }}
```
This is indeed better, since at least we don't really on `ps.form_group_row` anymore, but this could still be improved a bit more so that we only rely on a single `form_row(categoryForm)`
If the form contains some very specific custom parts to should be doable with a dedicated form theme

jolelievre commented on src/Core/Grid/Column/Type/DataColumn.php:
```diff
@@ -55,10 +55,12 @@ protected function configureOptions(OptionsResolver $resolver)
             ])
             ->setDefaults([
                 'clickable' => true,
+                'raw' => false,
```
This option doesn't seem used, so you should revert this modification I think

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -857,6 +878,8 @@ private function getCategoryToolbarButtons(Request $request, int $categoryId): a
      */
     private function getErrorMessages()
     {
+        $iniConfig = $this->get('prestashop.core.configuration.ini_configuration');
+
```
```suggestion
```
This is not used

jolelievre commented on src/Core/Grid/Definition/Factory/CategoryGridDefinitionFactory.php:
```diff
@@ -129,7 +128,7 @@ protected function getColumns()
             )
             ->add(
-                (new HtmlColumn('description'))
+                (new DataColumn('description'))
```
Why change this? Since the description can contain HTML code will it not be rendered weirdly with HTML tags and all? The advantage is that it will limit the size of displayed content (since the column includes a default limitation via `max_displayed_characters` but the content itself should be stripped from tags before it's sliced no? Which doesn't seem to be done anywhere for now, so it will display partially split HTML content

JevgenijVisockij replied:
I think it was accidentally changed after merge, since I added HtmlColumn in another PR. 
Fixed.

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -52,6 +53,8 @@
  */
 abstract class AbstractCategoryType extends TranslatorAwareType
 {
+    public const MAX_TITLE_LENGTH = 128;
```
This should be placed in a dedicated `CategorySettings` class so that it can be used in other code without relying on the form type itself

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -67,51 +70,80 @@ abstract class AbstractCategoryType extends TranslatorAwareType
+        if (isset($options['id_category'])) {
+            $disableMenuThumbnailsUpload = $options['disable_menu_thumbnails_upload'];
+        }
+        $genericCharactersHint = $this->trans('Invalid characters:', 'Admin.Global') . ' <>;=#{}';
```
```suggestion
        $genericCharactersHint = $this->trans('Invalid characters: %s', 'Admin.Notifications.Info', [TypedRegexValidator::CATALOG_CHARS]);
```

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -67,51 +70,80 @@ abstract class AbstractCategoryType extends TranslatorAwareType
                 'options' => [
+                    'attr' => [
+                        'maxlength' => static::MAX_TITLE_LENGTH,
+                    ],
```
I can't comment the lines below but I think you could replace most of the regexp constraints in this form type with `TypedRegex`

JevgenijVisockij replied:
Yeah you are right, replaced with TypedRegex

jolelievre commented on tests/Unit/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandlerTest.php:
```diff
@@ -0,0 +1,73 @@
+
+namespace PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataHandler;
+
+function scandir($directory, $sorting_order = null, $context = null)
```
This is far from ideal 😅 But I don't know how to improve this, usually I would recommend injecting a service to handle this listing so that you can mock and replace it, but this whole way of handling the max thumbnails seem very unstable and wrongly designed anyway So I'm not sure it's worth trying to improve this, it should be refactored completely I guess but it's not the purpose here

JevgenijVisockij committed:
Separating root category and non-root category handlers and commands

JevgenijVisockij committed:
Moved functions to their original place

JevgenijVisockij committed:
Changed back to HTML column

JevgenijVisockij committed:
Apply suggestions from code review

Removed uneeded code line and changing translation string

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Added category settings

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Type/CategoryThumbnailType.php:
```diff
@@ -0,0 +1,68 @@
+    {
+        return 'category_thumbnail';
+    }
+
```
```suggestion
```

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Type/CategoryMenuThumbnailsType.php:
```diff
@@ -0,0 +1,72 @@
+    {
+        return 'category_menu_thumbnails';
+    }
+
```
```suggestion
```

zuk3975 replied:
fix your IDE :D

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Type/CategoryCoverImageType.php:
```diff
@@ -0,0 +1,68 @@
+    {
+        return 'category_cover_image';
+    }
+
```
```suggestion
```

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -88,27 +101,35 @@ public function __construct(
         if (isset($options['id_category'])) {
-            $disableMenuThumbnailsUpload = $options['disable_menu_thumbnails_upload'];
+            $categoryId = (int) $options['id_category'];
+            $editableCategory = $this->queryBus->handle(new GetCategoryForEditing((int) $options['id_category']));
```
The query shouldn't be executed inside the abstract type but in the `CategoryFormDataProvider` (actually I believe it already is). So the data you're missing here should either be passed as data directly via the provider Or via options if you want to keep this separated from the data

If you need some "advanced options" based on the result of the query you can rely on the `FormOptionsProviderInterface` which has a similar behaviour as the `FormDataProviderInterface` except it builds an array of options that is passed to the form type when it's built

You can see an example here https://github.com/PrestaShop/PrestaShop/blob/3cd5ea8d86ab4578f914648cf423ad885d47c6e3/src/Core/Form/IdentifiableObject/OptionProvider/ProductFormOptionsProvider.php#L38 and you can then configure it in the form builder service definition:
https://github.com/PrestaShop/PrestaShop/blob/28cb9109c2027b4f8061548623c76cccb60e91eb/src/PrestaShopBundle/Resources/config/services/core/form/form_builder.yml#L202

The good thing is that the `getOptions` method also contains the `$data` generated via the data provider so you can use some of the data to generate your options

jolelievre commented on src/PrestaShopBundle/Form/Admin/Type/CategoryCoverImageType.php:
```diff
@@ -0,0 +1,68 @@
+    public function buildView(FormView $view, FormInterface $form, array $options)
+    {
+        parent::buildView($view, $form, $options);
+        $view->vars['download_url'] = $options['download_url'];
```
I don't know if you were aware of it but we have introduced a `download_url` on all `FileType` already via the `DownloadFileExtension` so it should already work and be handled via the form UI kit theme. Since you have a custom form theme for those elements you won't be able to use the generic form theme, but at least the option should already be handled and passed down to the view so you can simplify these types a bit

JevgenijVisockij replied:
It complains about not having download_url if I don't include it, even though i am extending FileType. 

I guess i could try to redefine downloadable_file to extend my new custom type but I don't know if that makes sense.

```
  form.extension.image_with_preview_downloadable_file:
    class: 'PrestaShopBundle\Form\Admin\Extension\DownloadFileExtension'
    tags:
      - { name: form.type_extension, extended_type: Symfony\Component\Form\Extension\Core\Type\FileType }
```

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/FormTheme/category.html.twig:
```diff
@@ -0,0 +1,29 @@
+{% extends '@PrestaShop/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig' %}
+
+{% block category_cover_image_widget %}
+  {% include '@PrestaShop/Admin/Sell/Catalog/Categories/Blocks/cover_image.html.twig' %}
```
I think you could move the content of those blocks back into this form theme file so that they are all in the same place

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/menu_thumbnail_images.html.twig:
```diff
@@ -24,14 +24,14 @@
+  {% if preview_images is defined and preview_images is not empty %}
     <div>
-      {% for menuThumbnail in menuThumbnailImages %}
+      {% for menuThumbnail in preview_images %}
```
I wonder if there would be a way to simplify these three form types into a single one, although I'm not sure it's possible but here are my suggestions:
- menu_thumbnail_images and cover_image seem to be very similar except for the fact the the first one has a list of images, but if you had a component relying always on a list it could also work for cover_image only it would always have an array of one element
- cover_image and category_thumbnail_image also seem quite similar, mostly the difference id that cover_image handles a delete method, but if you had a `can_be_deleted` option, then the display of the delete buttons could be conditioned

So I was thinking maybe (bu you'll know better than me) we could have a single form type that accepts a list of images and can option out the delete button in which case one form type and one template would be enough. But if it gets too complex it may become counterproductive, so let me know what you think about it

JevgenijVisockij replied:
Added those 3 together into 1

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -177,22 +198,26 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'label' => $this->trans('Category cover image', 'Admin.Catalog.Feature'),
                 'help' => $this->trans('This is the cover image for your category: it will be displayed on the category\'s page. The description will appear in its top-left corner.', 'Admin.Catalog.Help'),
                 'required' => false,
+                'preview_image' => $coverImage,
```
Regardless of the proposition I made upper, maybe it would be simpler to pass the thumbnail URLs as the data so you wouldn't even need to rely on the `FormOptionsProviderInterface` This is what is done in other form types that use `ImagePreviewType`

But the main difference is that `ImagePreviewType` is used purely as a display item, whereas you also need to handle the upload of files so not sure the string data of a URL is really compatible with the FileType

One last possibility would be to use two form types:
- one for preview that in which the URL data is inserted for display
- one for file upload that is only used on submit

JevgenijVisockij committed:
Fixed token for cover image

JevgenijVisockij committed:
Moved all category images to single type

JevgenijVisockij committed:
Moved category image to image with preview

JevgenijVisockij committed:
Fixed some issues with root category by aligning it closer with category

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php

JevgenijVisockij committed:
Removing unneeded use

JevgenijVisockij committed:
Replacing regex with typed Regex

JevgenijVisockij committed:
Fixing phpunit test

JevgenijVisockij committed:
reverting some unneded changes

JevgenijVisockij committed:
CS fixer fix

FabienPapet commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -200,4 +226,25 @@ private function uploadImages(
+     */
+    private function getAvailableKeys(int $categoryId): array
+    {
+        $files = scandir(_PS_CAT_IMG_DIR_, SCANDIR_SORT_NONE);
```
- We should not use constants but DI instead.

- Maybe use the Symfony finder for this, it has a stable API and has better error management support.

FabienPapet commented on src/Core/Form/IdentifiableObject/DataHandler/RootCategoryFormDataHandler.php:
```diff
@@ -157,35 +185,59 @@ private function createEditRootCategoryCommand($rootCategoryId, array $data)
+     */
+    private function getAvailableKeys(int $categoryId): array
+    {
+        $files = scandir(_PS_CAT_IMG_DIR_, SCANDIR_SORT_NONE);
```
same here, use DI and finder, maybe this code can be refactored, looks like it has been copy pasted

JevgenijVisockij replied:
Thanks! Refactored the code

FabienPapet commented on src/Core/Form/IdentifiableObject/DataProvider/CategoryFormDataProvider.php:
```diff
@@ -27,9 +27,11 @@
 namespace PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataProvider;
 
 use PrestaShop\PrestaShop\Adapter\Group\GroupDataProvider;
+use PrestaShop\PrestaShop\Adapter\Shop\Url\CategoryProvider;
```
Core should not depend on adapter, so this dependency cannot be injected here.

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/form/form_data_provider.yml:
```diff
@@ -46,6 +46,8 @@ services:
       - '@=service("prestashop.adapter.shop.context").getContextListShopID()'
       - '@=service("prestashop.adapter.legacy.context").getContext().shop.getCategory()'
       - '@prestashop.adapter.group.group_data_provider'
+      - '@prestashop.adapter.shop.url.category_provider'
```
As i said in another comment, the core cannot depend on the legacy here

JevgenijVisockij committed:
Fixed category not working

JevgenijVisockij committed:
Removing unneeded line

JevgenijVisockij committed:
Refactoring avialbe keys

JevgenijVisockij committed:
Removing unneeded include

FabienPapet commented on src/Core/Category/Provider/MenuThumbnailAvailableKeyProvider.php:
```diff
@@ -0,0 +1,58 @@
+    public function getAvailableKeys(int $categoryId): array
+    {
+        $finder = new Finder();
+        $finder->files()->in(_PS_CAT_IMG_DIR_);
```
Still a constant used here

FabienPapet commented on src/Core/Shop/Url/CategoryProvider.php:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * @param Link $link
+     */
+    public function __construct(Link $link)
```
This is a legacy class inside the core :/ If you want to do this, create an interface inside the legacy, and implement it inside the adapter namespace. It will help refactoring.

JevgenijVisockij replied:
You mentioned i shoudn't use adapter classes in Core, so I thought according to this ADR
https://github.com/PrestaShop/ADR/blob/master/0014-drop-the-no-use-of-legacy-in-new-code-restriction.md
Copy the implementation from Adapter to Core, I guess should also change the CategoryProvider in adapter class to use in core, but isn't the whole point of ADR that we can use legacy in core?

FabienPapet replied:
The main point here is that the class has been copied from the Adapter namespace into the Core namespace but was not deprecated nor removed from the Adapter namespace. Which leads of having the same file twice, one into the `Adapter` and one into the `Core`. And removing this class is a BC break. SO there's no need to create this file into the Core NS

From my totally subjective POV, ADR 14 is a non-sense and goes against migrations principles that you should not add any legacy code.

JevgenijVisockij committed:
Fixed issue with thumbnail warning not showing up

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij committed:
Adding constant to the yml

JevgenijVisockij committed:
Fixed add category page not working

zuk3975 commented:
conflcits, stan failing

l-delin commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -251,7 +347,21 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         if ($this->multistoreFeature->isUsed()) {
-            $builder->add('shop_association', ShopChoiceTreeType::class);
+            $builder->add('shop_association', ShopChoiceTreeType::class, [
+                'label' => $this->trans('Shop association', 'Admin.Global'),
```
```suggestion
                'label' => $this->trans('Store association', 'Admin.Global'),
```

JevgenijVisockij committed:
Adding raw option to data column

# Conflicts:
#	src/Core/Grid/Column/Type/DataColumn.php
#	src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/Content/data.html.twig

# Conflicts:
#	src/Core/Grid/Column/Type/DataColumn.php

JevgenijVisockij committed:
Fixed thumbnail error not working correctly

# Conflicts:
#	src/PrestaShopBundle/Resources/config/services/core/category.yml

JevgenijVisockij committed:
Fixed DataColumn

JevgenijVisockij committed:
Fixing phpstan issue

JevgenijVisockij committed:
Adding behat tests

JevgenijVisockij committed:
Adding behat tests

JevgenijVisockij committed:
Adding behat tests

zuk3975 commented on src/Core/Category/Provider/MenuThumbnailAvailableKeyProvider.php:
```diff
@@ -0,0 +1,68 @@
+use PrestaShop\PrestaShop\Core\Domain\Category\ValueObject\MenuThumbnailId;
+use Symfony\Component\Finder\Finder;
+
+class MenuThumbnailAvailableKeyProvider
```
could you add a comment above briefly explaining what this class does?
Its quite unique, would be nice to know some context.

jolelievre replied:
Yep I have no clue what this is used for 😅 I only understood it's some legacy code moved which is already a good thing I guess 😛

JevgenijVisockij replied:
Added it

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -211,36 +278,65 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                            'type' => TypedRegex::TYPE_LINK_REWRITE,
+                        ]),
+                        new Length([
+                            'max' => CategorySettings::MAX_TITLE_LENGTH,
```
```suggestion
                            'max' => CategorySettings::MAX_LINK_REWRITE_LENGTH,
```

jolelievre replied:
I even think it should be in `SeoSettings` considering it seems to contain settings about Category SEO related fields already (the expected value is 128 by the way)

zuk3975 commented on src/Core/Domain/Category/CategorySettings.php:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * Bellow constants define maximum allowed length of category properties
+     */
+    public const MAX_TITLE_LENGTH = 255;
```
As I understand title = name and I see from Category object model that name max length is 128.
Also after you fix this, you should also use this constant there in Category ObjectModel L 118
`            'name' => ['type' => self::TYPE_STRING, 'lang' => true, 'validate' => 'isCatalogName', 'required' => true, 'size' => CategorySettings::MAX_TITLE_LENGTH],
`

jolelievre replied:
```suggestion
    public const MAX_TITLE_LENGTH = 128;
```
Indeed the correct value is 128 so far

JevgenijVisockij replied:
Corrected link_rewrite and name to use correct lengths

zuk3975 commented on src/Core/Domain/Category/SeoSettings.php:
```diff
@@ -50,4 +50,9 @@ final class SeoSettings
+    /**
+     * Maximum length of SEO keywords (value is constrained by database)
+     */
+    public const MAX_KEYWORDS_LENGTH = 255;
```
same for these consts, make sure they have correct values based on object model and use these consts there as well

jolelievre replied:
This one seems correct based on 
https://github.com/PrestaShop/PrestaShop/blob/60756529c8aa7343c820db65b80767316948a49d/classes/Category.php#L134

But indeed this const should now replace the hard-coded value in `CategoryCore`

jolelievre replied:
Can you please replace in `classes/Category` the hard-coded values with the ones you added here? And same for the one in `CategorySettings` ideally

zuk3975 commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -101,17 +122,27 @@ public function create(array $data)
+        $categoryId = new CategoryId($categoryId);
 
-        $categoryId = new CategoryId((int) $categoryId);
+        /**
+         * In some cases in form menu_thumbnail_images can be disabled so value won't get here.
+         */
```
style preferences... just a suggestion
```suggestion
        // In some cases in form menu_thumbnail_images can be disabled so value won't get here.
```

zuk3975 replied:
I remember cs fixer complaining about smthing similar some time ago in one of my PR's :man_shrugging:

zuk3975 commented on src/Core/Form/IdentifiableObject/DataHandler/RootCategoryFormDataHandler.php:
```diff
@@ -99,17 +120,29 @@ public function create(array $data)
         $this->commandBus->handle($command);
+        $categoryId = new CategoryId($categoryId);
+
+        /**
+         * In some cases in form menu_thumbnail_images can be disabled so value won't get here.
+         */
```
same as above
```suggestion
        // In some cases in form menu_thumbnail_images can be disabled so value won't get here.
```

zuk3975 reviewed (changes requested):
couple comments regarding constraints

jolelievre commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -58,39 +64,54 @@ final class CategoryFormDataHandler implements FormDataHandlerInterface
      */
     public function create(array $data)
     {
+        if (!isset($data['menu_thumbnail_images']) && count($data['menu_thumbnail_images']) > count(MenuThumbnailId::ALLOWED_ID_VALUES)) {
+            throw new MenuThumbnailsLimitException('Maximum number of menu thumbnails exceeded for new category');
+        }
```
Seems like this check is very related to the Category business domain (or at least technical domain), so this check should be done in the handler. It's too much responsibility to ask the form data handler to do this.

The fact that `MenuThumbnailAvailableKeyProvider` is injected here is already a sign that too many things are done here (it's not the only one but that's another question 😅)

jolelievre commented on src/Core/Form/IdentifiableObject/DataHandler/RootCategoryFormDataHandler.php:
```diff
@@ -58,39 +64,54 @@ final class RootCategoryFormDataHandler implements FormDataHandlerInterface
      */
     public function create(array $data)
     {
+        if (!isset($data['menu_thumbnail_images']) && count($data['menu_thumbnail_images']) > count(MenuThumbnailId::ALLOWED_ID_VALUES)) {
+            throw new MenuThumbnailsLimitException('Maximum number of menu thumbnails exceeded for new category');
+        }
```
Same suggestion here, this should be done by the command handler

jolelievre commented on src/Core/Form/IdentifiableObject/DataHandler/RootCategoryFormDataHandler.php:
```diff
@@ -99,17 +120,29 @@ public function create(array $data)
     {
-        $command = $this->createEditRootCategoryCommand($rootCategoryId, $data);
+        $categoryId = (int) $categoryId;
+        $availableKeys = $this->menuThumbnailAvailableKeyProvider->getAvailableKeys($categoryId);
+
+        if (isset($data['menu_thumbnail_images']) && count($data['menu_thumbnail_images']) > count($availableKeys)) {
+            throw new MenuThumbnailsLimitException(sprintf('The maximum number of menu thumbnails has been reached for the %d category', $categoryId));
+        }
```
Also in the command handler

JevgenijVisockij replied:
Moved the whole image handling inside the Command Handlers

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/CategoryFormDataProvider.php:
```diff
@@ -82,6 +100,53 @@ public function getData($categoryId)
+        if ($categoryId) {
+            $categoryId = (int) $categoryId;
+            $categoryUrl = $this->categoryProvider->getUrl($categoryId, '{friendly-url}');
+            $editableCategory = $this->queryBus->handle(new GetCategoryForEditing($categoryId));
```
```suggestion
```
You don't need to execute this again, you already did it on line 83

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/CategoryFormDataProvider.php:
```diff
@@ -82,6 +100,53 @@ public function getData($categoryId)
 
+        $coverImages = $thumbnailImages = $menuThumbnailImagesEdited = null;
+        $disableMenuThumbnailsUpload = false;
+        if ($categoryId) {
```
You're not supposed to check this, `getData` is only used bor edition and the ID MUST be present

JevgenijVisockij replied:
Changed it so no longer checks for ID

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -363,7 +374,7 @@ public function editRootAction($categoryId, Request $request)
 
         try {
-            $rootCategoryForm = $rootCategoryFormBuilder->getFormFor((int) $categoryId);
+            $rootCategoryForm = $rootCategoryFormBuilder->getFormFor((int) $categoryId, []);
```
```suggestion
            $rootCategoryForm = $rootCategoryFormBuilder->getFormFor((int) $categoryId);
```
If you don't need to override the data then you don't need this second parameter (default value is already `[]`)

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -135,19 +165,56 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'required' => false,
+                'can_be_deleted' => true,
+                'show_size' => true,
+                'csrf_delete_token' => 'delete-cover-image',
```
```suggestion
                'csrf_delete_token_id' => 'delete-cover-image',
```
Otherwise it's confusing, you don't specify the token itself here but the ID/seed/key used to generate it in the twig file

JevgenijVisockij replied:
Good point, changed it to id

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/cover_image.html.twig:
```diff
@@ -24,13 +24,13 @@
 
 {% block category_cover_image %}
-  {% if coverImage is defined and coverImage is not null %}
+  {% if preview_image is defined and preview_image is not null %}
```
Is this twig file still needed with all the refacto you made based on `image_with_preview_widget` ?
Same question for:
- Categories/Blocks/menu_thumbnail_images.html.twig
- Categories/Blocks/thumbnail_image.html.twig

These three files used to be included in Categories/Blocks/form.html.twig but not anymore, if we can remove them all it's for the best!

JevgenijVisockij replied:
Yeah you are right they are no longer needed, removed them.

jolelievre reviewed (commented):
Thanks @JevgenijVisockij 

you must have struggled so much to handle this weird thumbnail management 😅
Nice work anyway!

FabienPapet commented on src/Core/Category/Provider/MenuThumbnailAvailableKeyProvider.php:
```diff
@@ -0,0 +1,68 @@
+
+        $usedKeys = [];
+
+        foreach ($finder as $file) {
+            $matches = [];
+
+            if (preg_match('/^' . $categoryId . '-([0-9])?_thumb.jpg/i', $file->getFilename(), $matches) === 1) {
+                $usedKeys[] = (int) $matches[1];
+            }
+        }
```
can you use the name filter of the finder for this ? 
https://symfony.com/doc/current/components/finder.html#file-name

JevgenijVisockij replied:
But Finder is already used here and preg match nicely finds ID, which exactly what I need.

FabienPapet commented on src/Core/Form/IdentifiableObject/DataProvider/CategoryFormDataProvider.php:
```diff
@@ -56,22 +58,38 @@ final class CategoryFormDataProvider implements FormDataProviderInterface
-        GroupDataProvider $groupDataProvider
+        GroupDataProvider $groupDataProvider,
+        CategoryProvider $categoryProvider,
+        Router $router
```
use UrlGeneratorInterface from symfony instead

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:
```diff
@@ -287,17 +287,19 @@ services:
     arguments:
       $customerGroupChoices: '@=service("prestashop.core.form.choice_provider.group_by_id").getChoices()'
       $multistoreFeature: '@prestashop.adapter.feature.multistore'
+      $router: '@prestashop.router'
```
use the default router by symfony

FabienPapet reviewed (dismissed):
Just a requested change to replace prestashop.router usages by symfony built it router

JevgenijVisockij committed:
Updated rewrite and category title length to be correct

JevgenijVisockij committed:
Apply suggestions from code review

Updating comment style

Co-authored-by: Julius Žukauskas <31609858+zuk3975@users.noreply.github.com>

JevgenijVisockij committed:
Moving image handling inside overall category handler

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataProvider/CategoryFormDataProvider.php

Removing unneeded line

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php

Removing unneeded param

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Changing csrf token to have id in its name for clarity

JevgenijVisockij committed:
Fixed issue with SErp not working properly

JevgenijVisockij committed:
Fixed serp not working properly on root category

JevgenijVisockij commented on admin-dev/themes/new-theme/js/pages/category/index.ts:
```diff
@@ -95,11 +95,11 @@ $(() => {
+      defaultDescription: 'textarea[name*="category[description]"]',
+      watchedDescription: 'textarea[name*="category[meta_description]"]',
+      watchedMetaUrl: 'input[name*="category[link_rewrite]"]',
       multiLanguageInput: `${translatorInput.localeInputSelector}:not(.d-none)`,
```
Before SERP didn't work properly in root category since all the names are like `root_category[link_rewrite][1]` and `name^` finds only inputs which start with `category[link_rewrite]`. `name*` finds any inputs with `category[link_rewrite]` which includes `root_category[link_rewrite]`

JevgenijVisockij committed:
Removed no longer needed files

JevgenijVisockij committed:
Changed to use symfony router

JevgenijVisockij committed:
Replacing rest of the routers to use symfony default

JevgenijVisockij committed:
Replacing rest of the routers to use symfony default

JevgenijVisockij committed:
Replacing rest of the routers to use symfony default

JevgenijVisockij committed:
Replacing rest of the routers to use symfony default

JevgenijVisockij committed:
Replacing rest of the routers to use symfony default

JevgenijVisockij committed:
Replacing rest of the routers to use symfony default

JevgenijVisockij committed:
Replacing rest of the routers to use symfony default

JevgenijVisockij committed:
Replacing rest of the routers to use symfony default

JevgenijVisockij committed:
Cs fixer

JevgenijVisockij committed:
Cs fixer

jolelievre commented on src/Core/Category/Provider/MenuThumbnailAvailableKeyProvider.php:
```diff
@@ -0,0 +1,73 @@
+ * Category can only have 3 thumbnails (0,1,2).
+ * We check if thumbnails with those id's exist, if they do they are no longer available.
+ */
+class MenuThumbnailAvailableKeyProvider
```
Thanks for the definition it's much clearer now, to make it perfect I think this could be unit tested By relying on some dedicated folders in `tests/Resources/category_menu_thumbnails` used for tests, this would show actual examples of what is expected from this class

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/CategoryFormDataProvider.php:
```diff
@@ -82,6 +100,47 @@ public function getData($categoryId)
+        $categoryUrl = $this->categoryProvider->getUrl($categoryId, '{friendly-url}');
+        $coverImage = $editableCategory->getCoverImage();
+        if ($coverImage) {
+            $coverImages[] = [
```
Are you sure you can consider `$coverImages` as an array and push a value in it? Because its initial value is `null`

Also I'm not sure you need an array of coverImages here since there can be only one, so maybe this could be enough:
```suggestion
            $coverImage = [
```
It's simply a flatten array for one image

jolelievre replied:
Ah unless you need to always provide a list of images because of the form type used for these three elements? Then I understand but then maybe `$coverImages` should be initialized with empty array instead of null?

JevgenijVisockij replied:
Yeah exactly, I need to work with menu thumbnails(which can be multiple), so cover and thumbnail are arrays if they consist only of single element.

I like syntax $coverImages a bit more, because it's still an array, and I use $coverImage for an actual image which I add to this array(of course I can rename it to something else if needed)

jolelievre replied:
No no I get it better now, it's good this ay!

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/CategoryFormDataProvider.php:
```diff
@@ -94,6 +153,11 @@ public function getData($categoryId)
             'link_rewrite' => $editableCategory->getLinkRewrite(),
             'group_association' => $editableCategory->getGroupAssociationIds(),
             'shop_association' => $editableCategory->getShopAssociationIds(),
+            'cover_image' => $coverImages,
+            'thumbnail_image' => $thumbnailImages,
```
```suggestion
            'cover_image' => $coverImage,
            'thumbnail_image' => $thumbnailImage,
```

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/CategoryFormDataProvider.php:
```diff
@@ -82,6 +100,47 @@ public function getData($categoryId)
+        }
+        $thumbnailImage = $editableCategory->getThumbnailImage();
+        if ($thumbnailImage) {
+            $thumbnailImages[] =
```
Same suggestion as for `coverImage `

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/FormTheme/category.html.twig:
```diff
@@ -0,0 +1,62 @@
+    </small>
+{% endblock category_seo_preview_widget %}
+
+{% block image_with_preview_widget %}
```
Since `ImageWithPreviewType` seems to be a form type generic (it's not in the category namespace) even though for now it's only used in category, I think this block should be moved into the generic form theme `prestashop_ui_kit_base.html.twig` file (only the `image_with_preview_widget` block)

JevgenijVisockij replied:
Sure I guess it makes sense. But I didn't build it with any other use cases in mind other then category page specifically needs. So if somebody uses it outside of it, it can be a bit tricky.

jolelievre reviewed (commented):
Thanks @JevgenijVisockij 

some final comments but everything is almost done. The handler look much cleaner now it's great!

I wonder if we could improve the behat tests to add some tests regarding the image upload. It's not very easy to do but we managed somehow in the product tests.

Although this could be done in a dedicated PR if we don't want this one to explode. The uni tests for `MenuThumbnailAvailableKeyProvider` should probably be added now hile it's still fresh in your mind 😅

JevgenijVisockij committed:
Code review changes

JevgenijVisockij committed:
Added test for menu thumbnail keys

JevgenijVisockij committed:
Added test for categoryImageFinder

JevgenijVisockij committed:
Added another test case

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/FormTheme/category.html.twig:
```diff
@@ -0,0 +1,8 @@
+{% extends '@PrestaShop/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig' %}
```
Missing license header on this twig file (soon we'll have a CI checking those ^^)

jolelievre reviewed (commented):
Thanks @JevgenijVisockij 

seems like there's only a missing header and CS fixer to be run (red for now) and we're all good!

JevgenijVisockij committed:
Cs Fixer fix

JevgenijVisockij committed:
Adding license header

jolelievre reviewed (approved):
Thanks @JevgenijVisockij

boherm reviewed (approved):
LGTM, thanks @JevgenijVisockij!

HanaRebaiQA reviewed (commented):
Hello @JevgenijVisockij 

Thank you for your PR. I have tested it. It's OK no issues, but : 

1. There are failures on automatic tests which are already present on Nightly
**Tests** : https://github.com/HanaRebaiQA/testing_pr/actions/runs/4786800852/jobs/8511191392
**Nightly** : https://github.com/PrestaShop/PrestaShop/actions/runs/4781547937/jobs/8500115028#step:5:511
ping @PrestaShop/qa-automation I think it's not related to this PR. Do you confirm it ?


2. @JevgenijVisockij Did you check @FabienPapet ' request change ?

Waiting for your feedback!

Thank You!

JevgenijVisockij commented:
> Hello @JevgenijVisockij
> 
> Thank you for your PR. I have tested it. It's OK no issues, but :
> 
>     1. There are failures on automatic tests which are already present on Nightly
>        **Tests** : https://github.com/HanaRebaiQA/testing_pr/actions/runs/4786800852/jobs/8511191392
>        **Nightly** : https://github.com/PrestaShop/PrestaShop/actions/runs/4781547937/jobs/8500115028#step:5:511
>        ping @PrestaShop/qa-automation I think it's not related to this PR. Do you confirm it ?
> 
>     2. @JevgenijVisockij Did you check @FabienPapet ' request change ?
> 
> 
> Waiting for your feedback!
> 
> Thank You!

Yeah, I believe I addressed all the asked changes, unless I missed something @FabienPapet ?

nesrineabdmouleh commented:
Hello @HanaRebaiQA,

The errors are not related to the PR, the same result exist in the run of another PR.
I will fix the failures in develop branch.

Thank you all

HanaRebaiQA reviewed (approved):
Thanks @FabienPapet 

So, it's QA :heavy_check_mark:

jolelievre commented:
Thanks @JevgenijVisockij and congratulations!!
